[REF] cetmix_tower_server: Refactor variable handling with mixin

tendil opened the pull request:
This commit refactors the handling of variables in the `cetmix_tower_server` module by moving `variable_ids` and related logic to a reusable mixin. This enhances code maintainability and provides consistent behavior across multiple models.

Key Changes:
- Added `TowerVariableMixin` to centralize variable handling logic.
- Moved `variable_ids` field and related logic to the mixin.
- Implemented `has_available_variables` in the mixin to control visibility of the "Variables" field in UI.
- Updated `cx.tower.command`, `cx.tower.file`, and `cx.tower.file.template` models to inherit from `TowerVariableMixin`.
- Removed redundant logic from individual models to reduce duplication.
- Ensured the "Variables" field in the form view is displayed only when related variables are available and visible to users with "Manager" or "Root" roles.

Why this change?
This refactor simplifies the codebase, reduces redundancy, and ensures consistent handling of variables across models. It also aligns with the project's goal of better modularity and scalability.

Backward Compatibility:
- Existing Many2many relationships (`variable_ids`) remain intact. No data migration is required as table relations, column names, and logic for linking variables are preserved.

Task: 4167

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new field, `server_status`, in the command model to track server status post-command execution.
	- Added `variable_ids` field as a read-only many-to-many tag widget in multiple views, enhancing interactivity and user experience.

- **Bug Fixes**
	- Removed outdated computation methods for `variable_ids`, streamlining its management across different models.

- **Documentation**
	- Updated view definitions to reflect the addition of new fields and their configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ivs-cetmix commented:
/ocabot rebase

CetmixGitDrone commented:
Congratulations, PR rebased to [14.0-dev](https://github.com/cetmix/cetmix-tower/commits/14.0-dev).

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_mixin.py:
```diff
@@ -13,13 +13,33 @@ class TowerVariableMixin(models.AbstractModel):
     variable_value_ids = fields.One2many(
         string="Variable Values",
         comodel_name="cx.tower.variable.value",
+        inverse_name="server_id",
```
What if this field is not related to server?

tendil replied:
Outdated

tendil committed:
[IMP] cetmix_tower_server: Centralize and enhance variable management logic

Centralized and streamlined the logic for managing `variable_ids` in `CxTowerTemplateMixin` and related models (`CxTowerCommand`, `CxTowerFile`, and `CxTowerFileTemplate`):

- Introduced `_get_relation_params` in `CxTowerTemplateMixin` to dynamically extract `relation`, `column1`, and `column2` for `variable_ids` based on field metadata.
- Created `_get_depends_fields` as a customizable method to define dependent fields for each model:
  - `CxTowerCommand` uses ["code", "path"].
  - `CxTowerFile` uses ["code", "server_dir", "name"].
  - `CxTowerFileTemplate` uses ["code", "server_dir", "file_name"].
- Enhanced `_compute_variable_ids` in `CxTowerTemplateMixin` to dynamically compute `variable_ids` using the dependencies provided by `_get_depends_fields`.
- Updated model definitions (`CxTowerCommand`, `CxTowerFile`, and `CxTowerFileTemplate`) to:
  - Retain unique relation details (`relation`, `column1`, and `column2`) for `variable_ids`.
  - Delegate computation logic to the mixin for better maintainability.
- Added professional docstrings for all methods, ensuring clear documentation and adherence to professional standards.

Additional changes:
- Adjusted UI views to include a read-only field `variable_ids`:
  - Visible only to users in the `Manager` and `Root` groups.

Task ID: 4167

tendil commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -115,32 +115,38 @@ def _selection_action(self):
+    )
     variable_ids = fields.Many2many(
         comodel_name="cx.tower.variable",
         relation="cx_tower_command_variable_rel",
         column1="command_id",
         column2="variable_id",
-        string="Variables",
-        compute="_compute_variable_ids",
-        store=True,
     )
 
```
We don't need to inherit `cx.tower.variable.mixin` since we are working through `cx.tower.template.mixin`

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -115,32 +115,38 @@ def _selection_action(self):
+    server_status = fields.Selection(
+        selection=lambda self: self.env["cx.tower.server"]._selection_status(),
+        string="Server Status",
+        help=(
```
No need to use brackets `()` here.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -16,6 +16,100 @@ class CxTowerTemplateMixin(models.AbstractModel):
+        """
+        relation_params = self._get_relation_params()
+        for record in self:
+            depends_fields = record._get_depends_fields()
```
Can those fields be different for different records? If not then you need to move this out of the loop.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -16,6 +16,100 @@ class CxTowerTemplateMixin(models.AbstractModel):
+        relation_params = self._get_relation_params()
+        for record in self:
+            depends_fields = record._get_depends_fields()
+            if relation_params["relation"] and depends_fields:
```
"relation" is a mandatory key for m2m fields. So no need to check for it explicitly.

ivs-cetmix reviewed (commented):
No need to provide `column1` and `column2` keys explicitly for m2m fields.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command.py:
```diff
@@ -117,33 +117,36 @@ def _selection_action(self):
     variable_ids = fields.Many2many(
         comodel_name="cx.tower.variable",
         relation="cx_tower_command_variable_rel",
         column1="command_id",
```
We can drop "column1" and "column2" too. They will be created automatically.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file.py:
```diff
@@ -135,20 +135,26 @@ class CxTowerFile(models.Model):
         relation="cx_tower_file_variable_rel",
         column1="file_id",
```
We can drop "column1" and "column2" too. They will be created automatically.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file_template.py:
```diff
@@ -53,27 +57,35 @@ def _compute_file_count(self):
-
     variable_ids = fields.Many2many(
         comodel_name="cx.tower.variable",
         relation="cx_tower_file_template_variable_rel",
```
We can drop "relation", "column1" and "column2" too. They will be created automatically.

ivs-cetmix reviewed (approved):
Code review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Aldeigja reviewed (approved):
Functional LGTM

hoangtrann reviewed (approved):
Code LGTM

ivs-cetmix commented:
/ocabot merge patch

CetmixGitDrone commented:
Hey, thanks for contributing! Proceeding to merge this for you.
Prepared branch [14.0-dev-ocabot-merge-pr-160-by-ivs-cetmix-bump-patch](https://github.com/cetmix/cetmix-tower/commits/14.0-dev-ocabot-merge-pr-160-by-ivs-cetmix-bump-patch), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at bb0cd2df23e1f26fe6ea77d9bf940c0047169fa6. Thanks a lot for contributing to cetmix. ❤️
